File: AccuPay/Domain/Payroll/OvertimeRate.vb
Option Strict On

Public Class OvertimeRate
    Public Property BasePay As Rate
    Public Property Overtime As Rate
    Public Property NightDifferential As Rate
    Public Property NightDifferentialOvertime As Rate

    Public Property RestDay As Rate
    Public Property RestDayOvertime As Rate
    Public Property RestDayNightDifferential As Rate
    Public Property RestDayNightDifferentialOvertime As Rate

    Public Property SpecialHoliday As Rate
    Public Property SpecialHolidayOvertime As Rate
    Public Property SpecialHolidayNightDifferential As Rate
    Public Property SpecialHolidayNightDifferentialOvertime As Rate
    Public Property SpecialHolidayRestDay As Rate
    Public Property SpecialHolidayRestDayOvertime As Rate
    Public Property SpecialHolidayRestDayNightDifferential As Rate
    Public Property SpecialHolidayRestDayNightDifferentialOvertime As Rate

    Public Property RegularHoliday As Rate
    Public Property RegularHolidayOvertime As Rate
    Public Property RegularHolidayNightDifferential As Rate
    Public Property RegularHolidayNightDifferentialOvertime As Rate
    Public Property RegularHolidayRestDay As Rate
    Public Property RegularHolidayRestDayOvertime As Rate
    Public Property RegularHolidayRestDayNightDifferential As Rate
    Public Property RegularHolidayRestDayNightDifferentialOvertime As Rate

    Public Const BasePayDescription As String = "Base Pay"
    Public Const OvertimeDescription As String = "Overtime"
    Public Const NightDifferentialDescription As String = "Night Differential"
    Public Const NightDifferentialOvertimeDescription As String = "Night Differential OT"

    Public Const RestDayDescription As String = "Rest Day"
    Public Const RestDayOvertimeDescription As String = "Rest Day OT"
    Public Const RestDayNightDifferentialDescription As String = "Rest Day Night Differential"
    Public Const RestDayNightDifferentialOvertimeDescription As String = "Rest Day Night Differential OT"

    Public Const SpecialHolidayDescription As String = "Special Holiday"
    Public Const SpecialHolidayOvertimeDescription As String = "Special Holiday OT"
    Public Const SpecialHolidayNightDifferentialDescription As String = "Special Holiday Night Differential"
    Public Const SpecialHolidayNightDifferentialOvertimeDescription As String = "Special Holiday Night Differential OT"
    Public Const SpecialHolidayRestDayDescription As String = "Special Holiday Rest Day"
    Public Const SpecialHolidayRestDayOvertimeDescription As String = "Special Holiday Rest Day OT"
    Public Const SpecialHolidayRestDayNightDifferentialDescription As String = "Special Holiday Rest Day Night Differential"
    Public Const SpecialHolidayRestDayNightDifferentialOvertimeDescription As String = "Special Holiday Rest Day Night Differential OT"

    Public Const RegularHolidayDescription As String = "Regular Holiday"
    Public Const RegularHolidayOvertimeDescription As String = "Regular Holiday OT"
    Public Const RegularHolidayNightDifferentialDescription As String = "Regular Holiday Night Differential"
    Public Const RegularHolidayNightDifferentialOvertimeDescription As String = "Regular Holiday Night Differential OT"
    Public Const RegularHolidayRestDayDescription As String = "Regular Holiday Rest Day"
    Public Const RegularHolidayRestDayOvertimeDescription As String = "Regular Holiday Rest Day OT"
    Public Const RegularHolidayRestDayNightDifferentialDescription As String = "Regular Holiday Rest Day Night Differential"
    Public Const RegularHolidayRestDayNightDifferentialOvertimeDescription As String = "Regular Holiday Rest Day Night Differential OT"

    Private _overtimeRateList As List(Of Rate)

    Public ReadOnly Property OvertimeRateList() As List(Of Rate)
        Get
            If _overtimeRateList Is Nothing Then

                _overtimeRateList = New List(Of Rate)

                '_rateList.Add(Me.BasePay)
                _overtimeRateList.Add(Me.Overtime)
                _overtimeRateList.Add(Me.NightDifferential)
                _overtimeRateList.Add(Me.NightDifferentialOvertime)
                _overtimeRateList.Add(Me.RestDay)
                _overtimeRateList.Add(Me.RestDayOvertime)
                _overtimeRateList.Add(Me.RestDayNightDifferential)
                _overtimeRateList.Add(Me.RestDayNightDifferentialOvertime)

                _overtimeRateList.Add(Me.SpecialHoliday)
                _overtimeRateList.Add(Me.SpecialHolidayOvertime)
                _overtimeRateList.Add(Me.SpecialHolidayNightDifferential)
                _overtimeRateList.Add(Me.SpecialHolidayNightDifferentialOvertime)
                _overtimeRateList.Add(Me.SpecialHolidayRestDay)
                _overtimeRateList.Add(Me.SpecialHolidayRestDayOvertime)
                _overtimeRateList.Add(Me.SpecialHolidayRestDayNightDifferential)
                _overtimeRateList.Add(Me.SpecialHolidayRestDayNightDifferentialOvertime)

                _overtimeRateList.Add(Me.RegularHoliday)
                _overtimeRateList.Add(Me.RegularHolidayOvertime)
                _overtimeRateList.Add(Me.RegularHolidayNightDifferential)
                _overtimeRateList.Add(Me.RegularHolidayNightDifferentialOvertime)
                _overtimeRateList.Add(Me.RegularHolidayRestDay)
                _overtimeRateList.Add(Me.RegularHolidayRestDayOvertime)
                _overtimeRateList.Add(Me.RegularHolidayRestDayNightDifferential)
                _overtimeRateList.Add(Me.RegularHolidayRestDayNightDifferentialOvertime)

            End If

            Return _overtimeRateList
        End Get
    End Property

    Public ReadOnly Property RegularRateGroup() As RateGroup
        Get
            Return New RateGroup() With {
                .BasePay = Me.BasePay,
                .Overtime = Me.Overtime,
                .NightDifferential = Me.NightDifferential,
                .NightDifferentialOvertime = Me.NightDifferentialOvertime,
                .RestDay = Me.RestDay,
                .RestDayOvertime = Me.RestDayOvertime,
                .RestDayNightDifferential = Me.RestDayNightDifferential,
                .RestDayNightDifferentialOvertime = Me.RestDayNightDifferentialOvertime
            }
        End Get
    End Property

    Public ReadOnly Property SpecialHolidayRateGroup() As RateGroup
        Get
            Return New RateGroup() With {
                .BasePay = Me.SpecialHoliday,
                .Overtime = Me.SpecialHolidayOvertime,
                .NightDifferential = Me.SpecialHolidayNightDifferential,
                .NightDifferentialOvertime = Me.SpecialHolidayNightDifferentialOvertime,
                .RestDay = Me.SpecialHolidayRestDay,
                .RestDayOvertime = Me.SpecialHolidayRestDayOvertime,
                .RestDayNightDifferential = Me.SpecialHolidayRestDayNightDifferential,
                .RestDayNightDifferentialOvertime = Me.SpecialHolidayRestDayNightDifferentialOvertime
            }
        End Get
    End Property

    Public ReadOnly Property RegularHolidayRateGroup() As RateGroup
        Get
            Return New RateGroup() With {
                .BasePay = Me.RegularHoliday,
                .Overtime = Me.RegularHolidayOvertime,
                .NightDifferential = Me.RegularHolidayNightDifferential,
                .NightDifferentialOvertime = Me.RegularHolidayNightDifferentialOvertime,
                .RestDay = Me.RegularHolidayRestDay,
                .RestDayOvertime = Me.RegularHolidayRestDayOvertime,
                .RestDayNightDifferential = Me.RegularHolidayRestDayNightDifferential,
                .RestDayNightDifferentialOvertime = Me.RegularHolidayRestDayNightDifferentialOvertime
            }
        End Get
    End Property

    Public Sub New(basePay As Decimal,
        overtime As Decimal,
        nightDifferential As Decimal,
        nightDifferentialOvertime As Decimal,
        restDay As Decimal,
        restDayOvertime As Decimal,
        restDayNightDifferential As Decimal,
        restDayNightDifferentialOvertime As Decimal,
        specialHoliday As Decimal,
        specialHolidayOvertime As Decimal,
        specialHolidayNightDifferential As Decimal,
        specialHolidayNightDifferentialOvertime As Decimal,
        specialHolidayRestDay As Decimal,
        specialHolidayRestDayOvertime As Decimal,
        specialHolidayRestDayNightDifferential As Decimal,
        specialHolidayRestDayNightDifferentialOvertime As Decimal,
        regularHoliday As Decimal,
        regularHolidayOvertime As Decimal,
        regularHolidayNightDifferential As Decimal,
        regularHolidayNightDifferentialOvertime As Decimal,
        regularHolidayRestDay As Decimal,
        regularHolidayRestDayOvertime As Decimal,
        regularHolidayRestDayNightDifferential As Decimal,
        regularHolidayRestDayNightDifferentialOvertime As Decimal)

        Me.BasePay = New Rate("Base Pay", basePay)
        Me.Overtime = New Rate("Overtime", overtime)
        Me.NightDifferential = New Rate("Night Differential", nightDifferential, Me.BasePay)
        Me.NightDifferentialOvertime = New Rate("Night Differential OT", nightDifferentialOvertime)

        Me.RestDay = New Rate("Rest Day", restDay)
        Me.RestDayOvertime = New Rate("Rest Day OT", restDayOvertime)
        Me.RestDayNightDifferential = New Rate("Rest Day Night Differential", restDayNightDifferential, Me.RestDay)
        Me.RestDayNightDifferentialOvertime = New Rate("Rest Day Night Differential OT", restDayNightDifferentialOvertime)

        Me.SpecialHoliday = New Rate("Special Holiday", specialHoliday)
        Me.SpecialHolidayOvertime = New Rate("Special Holiday OT", specialHolidayOvertime)
        Me.SpecialHolidayNightDifferential = New Rate("Special Holiday Night Differential", specialHolidayNightDifferential, Me.SpecialHoliday)
        Me.SpecialHolidayNightDifferentialOvertime = New Rate("Special Holiday Night Differential OT", specialHolidayNightDifferentialOvertime)
        Me.SpecialHolidayRestDay = New Rate("Special Holiday Rest Day", specialHolidayRestDay)
        Me.SpecialHolidayRestDayOvertime = New Rate("Special Holiday Rest Day OT", specialHolidayRestDayOvertime)
        Me.SpecialHolidayRestDayNightDifferential = New Rate("Special Holiday Rest Day Night Differential", specialHolidayRestDayNightDifferential, Me.SpecialHolidayRestDay)
        Me.SpecialHolidayRestDayNightDifferentialOvertime = New Rate("Special Holiday Rest Day Night Differential OT", specialHolidayRestDayNightDifferentialOvertime)

        Me.RegularHoliday = New Rate("Regular Holiday", regularHoliday)
        Me.RegularHolidayOvertime = New Rate("Regular Holiday OT", regularHolidayOvertime)
        Me.RegularHolidayNightDifferential = New Rate("Regular Holiday Night Differential", regularHolidayNightDifferential, Me.RegularHoliday)
        Me.RegularHolidayNightDifferentialOvertime = New Rate("Regular Holiday Night Differential OT", regularHolidayNightDifferentialOvertime)
        Me.RegularHolidayRestDay = New Rate("Regular Holiday Rest Day", regularHolidayRestDay)
        Me.RegularHolidayRestDayOvertime = New Rate("Regular Holiday Rest Day OT", regularHolidayRestDayOvertime)
        Me.RegularHolidayRestDayNightDifferential = New Rate("Regular Holiday Rest Day Night Differential", regularHolidayRestDayNightDifferential, Me.RegularHolidayRestDay)
        Me.RegularHolidayRestDayNightDifferentialOvertime = New Rate("Regular Holiday Rest Day Night Differential OT", regularHolidayRestDayNightDifferentialOvertime)

    End Sub

    Public Class RateGroup
        Public Property BasePay As Rate
        Public Property Overtime As Rate
        Public Property NightDifferential As Rate
        Public Property NightDifferentialOvertime As Rate

        Public Property RestDay As Rate
        Public Property RestDayOvertime As Rate
        Public Property RestDayNightDifferential As Rate
        Public Property RestDayNightDifferentialOvertime As Rate

    End Class

End Class

File: Accupay.Data/ValueObjects/OvertimeRate.cs
using System.Collections.Generic;

namespace AccuPay.Data.ValueObjects
{
    public class OvertimeRate
    {
        public Rate BasePay { get; set; }
        public Rate Overtime { get; set; }
        public Rate NightDifferential { get; set; }
        public Rate NightDifferentialOvertime { get; set; }

        public Rate RestDay { get; set; }
        public Rate RestDayOvertime { get; set; }
        public Rate RestDayNightDifferential { get; set; }
        public Rate RestDayNightDifferentialOvertime { get; set; }

        public Rate SpecialHoliday { get; set; }
        public Rate SpecialHolidayOvertime { get; set; }
        public Rate SpecialHolidayNightDifferential { get; set; }
        public Rate SpecialHolidayNightDifferentialOvertime { get; set; }
        public Rate SpecialHolidayRestDay { get; set; }
        public Rate SpecialHolidayRestDayOvertime { get; set; }
        public Rate SpecialHolidayRestDayNightDifferential { get; set; }
        public Rate SpecialHolidayRestDayNightDifferentialOvertime { get; set; }

        public Rate RegularHoliday { get; set; }
        public Rate RegularHolidayOvertime { get; set; }
        public Rate RegularHolidayNightDifferential { get; set; }
        public Rate RegularHolidayNightDifferentialOvertime { get; set; }
        public Rate RegularHolidayRestDay { get; set; }
        public Rate RegularHolidayRestDayOvertime { get; set; }
        public Rate RegularHolidayRestDayNightDifferential { get; set; }
        public Rate RegularHolidayRestDayNightDifferentialOvertime { get; set; }

        public Rate DoubleHoliday { get; set; }
        public Rate DoubleHolidayOvertime { get; set; }
        public Rate DoubleHolidayNightDifferential { get; set; }
        public Rate DoubleHolidayNightDifferentialOvertime { get; set; }
        public Rate DoubleHolidayRestDay { get; set; }
        public Rate DoubleHolidayRestDayOvertime { get; set; }
        public Rate DoubleHolidayRestDayNightDifferential { get; set; }
        public Rate DoubleHolidayRestDayNightDifferentialOvertime { get; set; }

        public const string BasePayDescription = "Base Pay";
        public const string OvertimeDescription = "Overtime";
        public const string NightDifferentialDescription = "Night Differential";
        public const string NightDifferentialOvertimeDescription = "Night Differential OT";

        public const string RestDayDescription = "Rest Day";
        public const string RestDayOvertimeDescription = "Rest Day OT";
        public const string RestDayNightDifferentialDescription = "Rest Day Night Differential";
        public const string RestDayNightDifferentialOvertimeDescription = "Rest Day Night Differential OT";

        public const string SpecialHolidayDescription = "Special Holiday";
        public const string SpecialHolidayOvertimeDescription = "Special Holiday OT";
        public const string SpecialHolidayNightDifferentialDescription = "Special Holiday Night Differential";
        public const string SpecialHolidayNightDifferentialOvertimeDescription = "Special Holiday Night Differential OT";
        public const string SpecialHolidayRestDayDescription = "Special Holiday Rest Day";
        public const string SpecialHolidayRestDayOvertimeDescription = "Special Holiday Rest Day OT";
        public const string SpecialHolidayRestDayNightDifferentialDescription = "Special Holiday Rest Day Night Differential";
        public const string SpecialHolidayRestDayNightDifferentialOvertimeDescription = "Special Holiday Rest Day Night Differential OT";

        public const string RegularHolidayDescription = "Regular Holiday";
        public const string RegularHolidayOvertimeDescription = "Regular Holiday OT";
        public const string RegularHolidayNightDifferentialDescription = "Regular Holiday Night Differential";
        public const string RegularHolidayNightDifferentialOvertimeDescription = "Regular Holiday Night Differential OT";
        public const string RegularHolidayRestDayDescription = "Regular Holiday Rest Day";
        public const string RegularHolidayRestDayOvertimeDescription = "Regular Holiday Rest Day OT";
        public const string RegularHolidayRestDayNightDifferentialDescription = "Regular Holiday Rest Day Night Differential";
        public const string RegularHolidayRestDayNightDifferentialOvertimeDescription = "Regular Holiday Rest Day Night Differential OT";

        public const string DoubleHolidayDescription = "Double Holiday";
        public const string DoubleHolidayOvertimeDescription = "Double Holiday OT";
        public const string DoubleHolidayNightDifferentialDescription = "Double Holiday Night Differential";
        public const string DoubleHolidayNightDifferentialOvertimeDescription = "ReguDoublelar Holiday Night Differential OT";
        public const string DoubleHolidayRestDayDescription = "Double Holiday Rest Day";
        public const string DoubleHolidayRestDayOvertimeDescription = "Double Holiday Rest Day OT";
        public const string DoubleHolidayRestDayNightDifferentialDescription = "Double Holiday Rest Day Night Differential";
        public const string DoubleHolidayRestDayNightDifferentialOvertimeDescription = "Double Holiday Rest Day Night Differential OT";

        private List<Rate> _overtimeRateList;

        public List<Rate> OvertimeRateList
        {
            get
            {
                if (_overtimeRateList == null)
                {
                    _overtimeRateList = new List<Rate>();

                    // _rateList.Add(Me.BasePay)
                    _overtimeRateList.Add(this.Overtime);
                    _overtimeRateList.Add(this.NightDifferential);
                    _overtimeRateList.Add(this.NightDifferentialOvertime);
                    _overtimeRateList.Add(this.RestDay);
                    _overtimeRateList.Add(this.RestDayOvertime);
                    _overtimeRateList.Add(this.RestDayNightDifferential);
                    _overtimeRateList.Add(this.RestDayNightDifferentialOvertime);

                    _overtimeRateList.Add(this.SpecialHoliday);
                    _overtimeRateList.Add(this.SpecialHolidayOvertime);
                    _overtimeRateList.Add(this.SpecialHolidayNightDifferential);
                    _overtimeRateList.Add(this.SpecialHolidayNightDifferentialOvertime);
                    _overtimeRateList.Add(this.SpecialHolidayRestDay);
                    _overtimeRateList.Add(this.SpecialHolidayRestDayOvertime);
                    _overtimeRateList.Add(this.SpecialHolidayRestDayNightDifferential);
                    _overtimeRateList.Add(this.SpecialHolidayRestDayNightDifferentialOvertime);

                    _overtimeRateList.Add(this.RegularHoliday);
                    _overtimeRateList.Add(this.RegularHolidayOvertime);
                    _overtimeRateList.Add(this.RegularHolidayNightDifferential);
                    _overtimeRateList.Add(this.RegularHolidayNightDifferentialOvertime);
                    _overtimeRateList.Add(this.RegularHolidayRestDay);
                    _overtimeRateList.Add(this.RegularHolidayRestDayOvertime);
                    _overtimeRateList.Add(this.RegularHolidayRestDayNightDifferential);
                    _overtimeRateList.Add(this.RegularHolidayRestDayNightDifferentialOvertime);
                }

                return _overtimeRateList;
            }
        }

        public RateGroup RegularRateGroup
        {
            get
            {
                return new RateGroup()
                {
                    BasePay = this.BasePay,
                    Overtime = this.Overtime,
                    NightDifferential = this.NightDifferential,
                    NightDifferentialOvertime = this.NightDifferentialOvertime,
                    RestDay = this.RestDay,
                    RestDayOvertime = this.RestDayOvertime,
                    RestDayNightDifferential = this.RestDayNightDifferential,
                    RestDayNightDifferentialOvertime = this.RestDayNightDifferentialOvertime
                };
            }
        }

        public RateGroup SpecialHolidayRateGroup
        {
            get
            {
                return new RateGroup()
                {
                    BasePay = this.SpecialHoliday,
                    Overtime = this.SpecialHolidayOvertime,
                    NightDifferential = this.SpecialHolidayNightDifferential,
                    NightDifferentialOvertime = this.SpecialHolidayNightDifferentialOvertime,
                    RestDay = this.SpecialHolidayRestDay,
                    RestDayOvertime = this.SpecialHolidayRestDayOvertime,
                    RestDayNightDifferential = this.SpecialHolidayRestDayNightDifferential,
                    RestDayNightDifferentialOvertime = this.SpecialHolidayRestDayNightDifferentialOvertime
                };
            }
        }

        public RateGroup RegularHolidayRateGroup
        {
            get
            {
                return new RateGroup()
                {
                    BasePay = this.RegularHoliday,
                    Overtime = this.RegularHolidayOvertime,
                    NightDifferential = this.RegularHolidayNightDifferential,
                    NightDifferentialOvertime = this.RegularHolidayNightDifferentialOvertime,
                    RestDay = this.RegularHolidayRestDay,
                    RestDayOvertime = this.RegularHolidayRestDayOvertime,
                    RestDayNightDifferential = this.RegularHolidayRestDayNightDifferential,
                    RestDayNightDifferentialOvertime = this.RegularHolidayRestDayNightDifferentialOvertime
                };
            }
        }

        public RateGroup DoubleHolidayRateGroup
        {
            get
            {
                return new RateGroup()
                {
                    BasePay = this.DoubleHoliday,
                    Overtime = this.DoubleHolidayOvertime,
                    NightDifferential = this.DoubleHolidayNightDifferential,
                    NightDifferentialOvertime = this.DoubleHolidayNightDifferentialOvertime,
                    RestDay = this.DoubleHolidayRestDay,
                    RestDayOvertime = this.DoubleHolidayRestDayOvertime,
                    RestDayNightDifferential = this.DoubleHolidayRestDayNightDifferential,
                    RestDayNightDifferentialOvertime = this.DoubleHolidayRestDayNightDifferentialOvertime
                };
            }
        }

        public OvertimeRate(decimal basePay, decimal overtime, decimal nightDifferential, decimal nightDifferentialOvertime, decimal restDay, decimal restDayOvertime, decimal restDayNightDifferential, decimal restDayNightDifferentialOvertime, decimal specialHoliday, decimal specialHolidayOvertime, decimal specialHolidayNightDifferential, decimal specialHolidayNightDifferentialOvertime, decimal specialHolidayRestDay, decimal specialHolidayRestDayOvertime, decimal specialHolidayRestDayNightDifferential, decimal specialHolidayRestDayNightDifferentialOvertime, decimal regularHoliday, decimal regularHolidayOvertime, decimal regularHolidayNightDifferential, decimal regularHolidayNightDifferentialOvertime, decimal regularHolidayRestDay, decimal regularHolidayRestDayOvertime, decimal regularHolidayRestDayNightDifferential, decimal regularHolidayRestDayNightDifferentialOvertime, decimal doubleHoliday, decimal doubleHolidayOvertime, decimal doubleHolidayNightDifferential, decimal doubleHolidayNightDifferentialOvertime, decimal doubleHolidayRestDay, decimal doubleHolidayRestDayOvertime, decimal doubleHolidayRestDayNightDifferential, decimal doubleHolidayRestDayNightDifferentialOvertime)
        {
            this.BasePay = new Rate("Base Pay", basePay);
            this.Overtime = new Rate("Overtime", overtime);
            this.NightDifferential = new Rate("Night Differential", nightDifferential, this.BasePay);
            this.NightDifferentialOvertime = new Rate("Night Differential OT", nightDifferentialOvertime);

            this.RestDay = new Rate("Rest Day", restDay);
            this.RestDayOvertime = new Rate("Rest Day OT", restDayOvertime);
            this.RestDayNightDifferential = new Rate("Rest Day Night Differential", restDayNightDifferential, this.RestDay);
            this.RestDayNightDifferentialOvertime = new Rate("Rest Day Night Differential OT", restDayNightDifferentialOvertime);

            this.SpecialHoliday = new Rate("Special Holiday", specialHoliday);
            this.SpecialHolidayOvertime = new Rate("Special Holiday OT", specialHolidayOvertime);
            this.SpecialHolidayNightDifferential = new Rate("Special Holiday Night Differential", specialHolidayNightDifferential, this.SpecialHoliday);
            this.SpecialHolidayNightDifferentialOvertime = new Rate("Special Holiday Night Differential OT", specialHolidayNightDifferentialOvertime);
            this.SpecialHolidayRestDay = new Rate("Special Holiday Rest Day", specialHolidayRestDay);
            this.SpecialHolidayRestDayOvertime = new Rate("Special Holiday Rest Day OT", specialHolidayRestDayOvertime);
            this.SpecialHolidayRestDayNightDifferential = new Rate("Special Holiday Rest Day Night Differential", specialHolidayRestDayNightDifferential, this.SpecialHolidayRestDay);
            this.SpecialHolidayRestDayNightDifferentialOvertime = new Rate("Special Holiday Rest Day Night Differential OT", specialHolidayRestDayNightDifferentialOvertime);

            this.RegularHoliday = new Rate("Regular Holiday", regularHoliday);
            this.RegularHolidayOvertime = new Rate("Regular Holiday OT", regularHolidayOvertime);
            this.RegularHolidayNightDifferential = new Rate("Regular Holiday Night Differential", regularHolidayNightDifferential, this.RegularHoliday);
            this.RegularHolidayNightDifferentialOvertime = new Rate("Regular Holiday Night Differential OT", regularHolidayNightDifferentialOvertime);
            this.RegularHolidayRestDay = new Rate("Regular Holiday Rest Day", regularHolidayRestDay);
            this.RegularHolidayRestDayOvertime = new Rate("Regular Holiday Rest Day OT", regularHolidayRestDayOvertime);
            this.RegularHolidayRestDayNightDifferential = new Rate("Regular Holiday Rest Day Night Differential", regularHolidayRestDayNightDifferential, this.RegularHolidayRestDay);
            this.RegularHolidayRestDayNightDifferentialOvertime = new Rate("Regular Holiday Rest Day Night Differential OT", regularHolidayRestDayNightDifferentialOvertime);

            this.DoubleHoliday = new Rate("Double Holiday", doubleHoliday);
            this.DoubleHolidayOvertime = new Rate("Double Holiday OT", doubleHolidayOvertime);
            this.DoubleHolidayNightDifferential = new Rate("Double Holiday Night Differential", doubleHolidayNightDifferential, this.DoubleHoliday);
            this.DoubleHolidayNightDifferentialOvertime = new Rate("Double Holiday Night Differential OT", doubleHolidayNightDifferentialOvertime);
            this.DoubleHolidayRestDay = new Rate("Double Holiday Rest Day", doubleHolidayRestDay);
            this.DoubleHolidayRestDayOvertime = new Rate("Double Holiday Rest Day OT", doubleHolidayRestDayOvertime);
            this.DoubleHolidayRestDayNightDifferential = new Rate("Double Holiday Rest Day Night Differential", doubleHolidayRestDayNightDifferential, this.DoubleHoliday);
            this.DoubleHolidayRestDayNightDifferentialOvertime = new Rate("Double Holiday Rest Day Night Differential OT", doubleHolidayRestDayNightDifferentialOvertime);

            this.DoubleHoliday = new Rate("Double Holiday", doubleHoliday);
            this.DoubleHolidayOvertime = new Rate("Double Holiday OT", doubleHolidayOvertime);
            this.DoubleHolidayNightDifferential = new Rate("Double Holiday Night Differential", doubleHolidayNightDifferential, this.DoubleHoliday);
            this.DoubleHolidayNightDifferentialOvertime = new Rate("Double Holiday Night Differential OT", doubleHolidayNightDifferentialOvertime);
            this.DoubleHolidayRestDay = new Rate("Double Holiday Rest Day", doubleHolidayRestDay);
            this.DoubleHolidayRestDayOvertime = new Rate("Double Holiday Rest Day OT", doubleHolidayRestDayOvertime);
            this.DoubleHolidayRestDayNightDifferential = new Rate("Double Holiday Rest Day Night Differential", doubleHolidayRestDayNightDifferential, this.DoubleHoliday);
            this.DoubleHolidayRestDayNightDifferentialOvertime = new Rate("Double Holiday Rest Day Night Differential OT", doubleHolidayRestDayNightDifferentialOvertime);
        }

        public class RateGroup
        {
            public Rate BasePay { get; set; }
            public Rate Overtime { get; set; }
            public Rate NightDifferential { get; set; }
            public Rate NightDifferentialOvertime { get; set; }

            public Rate RestDay { get; set; }
            public Rate RestDayOvertime { get; set; }
            public Rate RestDayNightDifferential { get; set; }
            public Rate RestDayNightDifferentialOvertime { get; set; }
        }
    }
}